bookwyrm/views/reading.py

```diff
@@ -83,7 +84,7 @@
             if current_status_shelfbook.shelf.identifier != desired_shelf.identifier:
                 current_status_shelfbook.delete()
             else:  # It already was on the shelf
-                return redirect("/")
+                return redirect(next_step)
```

Comment:
Ah yes, it seems serious.

A solution could be to encode/sign the value on the template, and decode it in the view 🤔 

Github's help page says:

> To guard against untrusted URL redirection, it is advisable to avoid putting user input directly into a redirect URL. Instead, maintain a list of authorized redirects on the server; then choose from that list based on the user input provided.
> https://codeql.github.com/codeql-query-help/csharp/cs-web-unvalidated-url-redirection/

Would it be more practical?

Reply:
Another solution would be:

- on form/page load, the path is stored along with a key (on Redis perhaps?)
- the key is included as the hidden input value
- on form submit, the view looks up the key in storage and redirects to the path linked to that key (if found, else `/`)
- then the view deletes the key/path from storage

Reply:
Would the `Referer` request header be an option maybe? I haven't fully tested it out, but adding a book to "want to read" (i.e. `POST reading-status/want/<book_id>`) sets the Referer to the book page from which the request originated.

Reply:
Yes, I used to use referer but removed it for security reasons. I think validating it is the right way! There might be something in Django's url res

Reply:
What would be the best validation strategy? Just validating for the domain name seems not enough, is there a way to check that the path is valid against `urls.py`, or a subset thereof?

Reply:
I was thinking that validating that the url is prefixed with the protocol and domain should be enough in this case, no? This would prevent phising potential phising attacks of redirecting the user to another site looking exactly the same. And since it's a client-side redirect, the usual authorization applies to whatever page the user is redirected to. But keep in mind that I'm no security expert!

Reply:
That seems reasonable to me; I thought there'd be a clever way to resolve the path using Djangos' urls but I haven't found it

Reply:
I would have loved a way to validated using urls.py, but in the meantime I've added a simple validator https://github.com/bookwyrm-social/bookwyrm/pull/2544#commits-pushed-5c92774

Would you prefer I change from using the hidden input to using the referer?

Reply:
Thank you! And yes, using the http header will be more flexible and robust